Installation issue: `openblas` does not have a correct `install_name` on macos

### Steps to reproduce the issue

_No response_

### Error message

`openblas` has an `install_name` of `libopenblas.0.dylib` which means that some consuming packages struggle to consume libraries linked against openblas.

For example, `py-numpy` links against `openblas`, but when `py-numpy` is used by `py-netcdf4`, `py-netcdf4` fails to build as numpy does not prefix `@rpath/`. 

```
$ otool -L /Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/py-numpy-1.26.4-47jedfkjvehucaeawa47zz5xm7vkogjd/lib/python3.11/site-packages/numpy/core/_multiarray_umath.cpython-311-darwin.so
/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/py-numpy-1.26.4-47jedfkjvehucaeawa47zz5xm7vkogjd/lib/python3.11/site-packages/numpy/core/_multiarray_umath.cpython-311-darwin.so:
	libopenblas.0.dylib (compatibility version 0.0.0, current version 0.3.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1345.100.2)
``` 

```
[...]
          cmd LC_RPATH
      cmdsize 160
         path /Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/openblas-0.3.27-oecxneqr7e2rdzlmv36r6wq4dn5aac2j/lib (offset 12)
Load command 26
```

```
$ spack install py-netcdf4
[...]

  Traceback (most recent call last):
    File "/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/py-numpy-1.26.4-47jedfkjvehucaeawa47zz5xm7vkogjd/lib/python3.11/site-packages/numpy/core/__init__.py", line 24, in <module>
      from . import multiarray
    File "/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/py-numpy-1.26.4-47jedfkjvehucaeawa47zz5xm7vkogjd/lib/python3.11/site-packages/numpy/core/multiarray.py", line 10, in <module>
      from . import overrides
    File "/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/py-numpy-1.26.4-47jedfkjvehucaeawa47zz5xm7vkogjd/lib/python3.11/site-packages/numpy/core/overrides.py", line 8, in <module>
      from numpy.core._multiarray_umath import (
  ImportError: dlopen(/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/py-numpy-1.26.4-47jedfkjvehucaeawa47zz5xm7vkogjd/lib/python3.11/site-packages/numpy/core/_multiarray_umath.cpython-311-darwin.so, 0x0002): Library not loaded: libopenblas.0.dylib
    Referenced from: <96301193-7C5B-311E-9A15-E84A444ECE1C> /Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/py-numpy-1.26.4-47jedfkjvehucaeawa47zz5xm7vkogjd/lib/python3.11/site-packages/numpy/core/_multiarray_umath.cpython-311-darwin.so
    Reason: tried: 'libopenblas.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibopenblas.0.dylib' (no such file), 'libopenblas.0.dylib' (no such file), '/Users/cmarsh/Documents/science/code/spack/stage/spack-stage-py-netcdf4-1.6.5-r52uf346tmcn7fmhko3phfo5vnztmt67/spack-src/libopenblas.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/cmarsh/Documents/science/code/spack/stage/spack-stage-py-netcdf4-1.6.5-r52uf346tmcn7fmhko3phfo5vnztmt67/spack-src/libopenblas.0.dylib' (no such file), '/Users/cmarsh/Documents/science/code/spack/stage/spack-stage-py-netcdf4-1.6.5-r52uf346tmcn7fmhko3phfo5vnztmt67/spack-src/libopenblas.0.dylib' (no such file)
```

/cc @adamjstewart 
Adam, this is the root of the build errors I reported in Slack yesterday.

I think the simplest solution is to ensure openblas has a sane install_name.

### Information on your system

```
spack debug report
* **Spack:** 0.23.0.dev0 (4240748ceac4eafd2e1b34dd98a93db41cbf5796)
* **Python:** 3.12.3
* **Platform:** darwin-sonoma-m1
* **Concretizer:** clingo
```

### Additional information

_No response_

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

Ya I was surprised too.

```
$ spack spec /oo2xo5qoyvsop3zyogpq3ngllgv7il2r
Input spec
--------------------------------
 -   /oo2xo5qoyvsop3zyogpq3ngllgv7il2r

Concretized
--------------------------------
[+]  openblas@0.3.27%apple-clang@15.0.0~bignuma~consistent_fpcsr+dynamic_dispatch+fortran~ilp64~ipo+locking+pic+shared build_system=cmake build_type=Release generator=make symbol_suffix=none threads=none arch=darwin-sonoma-m1
[+]      ^cmake@3.30.1%apple-clang@15.0.0~doc+ncurses+ownlibs build_system=generic build_type=Release patches=dbc3892 arch=darwin-sonoma-m1

$ otool -D libopenblas.0.3.dylib
libopenblas.0.3.dylib:
libopenblas.0.dylib
```

I can't get the makefile variant to build -- I think it's pikcing up homebrew's gfortran so I cannot test what its output looks like.
```
==> Error: ProcessError: Command exited with status 2:
    'make' '-j10' '-s' 'CC=/Users/cmarsh/Documents/science/code/spack/lib/spack/env/clang/clang' 'FC=/Users/cmarsh/Documents/science/code/spack/lib/spack/env/clang/gfortran' 'MAKE_NB_JOBS=0' 'ARCH=arm64' 'DYNAMIC_ARCH=1' 'DYNAMIC_OLDER=1' 'TARGET=GENERIC' 'DYNAMIC_ARCH=1' 'USE_LOCKING=1' 'USE_OPENMP=0' 'USE_THREAD=0' 'RANLIB=ranlib' 'all'

2 errors found in build log:
     17340    ld: warning: could not create compact unwind for _dlatm1_: does not use standard frame
     17341    ld: warning: could not create compact unwind for _dlatm7_: does not use standard frame
     17342    ld: warning: could not create compact unwind for _dlarnd_: does not use standard frame
     17343    dyld[31106]: Library not loaded: @rpath/libgfortran.5.dylib
     17344      Referenced from: <4DE0C938-0F1A-33E0-89CA-8C125666D516> /Users/cmarsh/Documents/science/code/spack/stage/spack-stage-openblas-0.3.27-asrysdsjcx6zr7hxrsf5igpep2xe7ty6/spack-src/utest/openblas_utest
     17345      Reason: tried: '/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/openblas-0.3.27-asrysdsjcx6zr7hxrsf5igpep2xe7ty6/lib/libgfortran.5.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/openblas-0.3.27-asrysdsjcx6zr7hxrsf5igpep2xe7ty6/lib/libgfortran.5.dylib' (no
              such file), '/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/openblas-0.3.27-asrysdsjcx6zr7hxrsf5igpep2xe7ty6/lib64/libgfortran.5.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/openblas-0.3.27-asrysdsjcx6zr7hxrsf5igpep2xe7ty6/lib64/libgfortran.5.dylib' (no s
              uch file), '/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/openblas-0.3.27-asrysdsjcx6zr7hxrsf5igpep2xe7ty6/lib/libgfortran.5.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/openblas-0.3.27-asrysdsjcx6zr7hxrsf5igpep2xe7ty6/lib/libgfortran.5.dylib' (no such f
              ile), '/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/openblas-0.3.27-asrysdsjcx6zr7hxrsf5igpep2xe7ty6/lib64/libgfortran.5.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/openblas-0.3.27-asrysdsjcx6zr7hxrsf5igpep2xe7ty6/lib64/libgfortran.5.dylib' (no such fi
              le)
  >> 17346    make[1]: *** [Makefile:81: run_test] Abort trap: 6
  >> 17347    make: *** [Makefile:168: tests] Error 2

```

--- a/var/spack/repos/builtin/packages/openblas/package.py
+++ b/var/spack/repos/builtin/packages/openblas/package.py
@@ -612,6 +612,8 @@ class CMakeBuilder(spack.build_systems.cmake.CMakeBuilder):
 
         return cmake_defs
 
+    # TODO: constrain this to @:0.3.27 once @0.3.28 is released released and confirmed fixed
+    # https://github.com/OpenMathLib/OpenBLAS/pull/4840
     @run_after("install", when="platform=darwin")
     def darwin_install_name(self):
         # The shared library is not installed correctly on Darwin; fix this

--- a/var/spack/repos/builtin/packages/openblas/package.py
+++ b/var/spack/repos/builtin/packages/openblas/package.py
@@ -611,3 +611,10 @@ class CMakeBuilder(spack.build_systems.cmake.CMakeBuilder):
             cmake_defs += [self.define("USE_OPENMP", "OFF"), self.define("USE_THREAD", "OFF")]
 
         return cmake_defs
+
+    @run_after("install")
+    @when("platform=darwin")
+    def darwin_install_name(self):
+        # The shared library is not installed correctly on Darwin; fix this
+        fix_darwin_install_name(self.prefix.lib)
+

--- a/var/spack/repos/builtin/packages/openblas/package.py
+++ b/var/spack/repos/builtin/packages/openblas/package.py
@@ -588,7 +588,12 @@ class MakefileBuilder(spack.build_systems.makefile.MakefileBuilder):
 
 class CMakeBuilder(spack.build_systems.cmake.CMakeBuilder):
     def cmake_args(self):
-        cmake_defs = [self.define("TARGET", "GENERIC")]
+        cmake_defs = [
+            self.define("TARGET", "GENERIC"),
+            # ensure MACOSX_RPATH is set
+            self.define("CMAKE_POLICY_DEFAULT_CMP0042", "NEW"),
+        ]
+
         if self.spec.satisfies("+dynamic_dispatch"):
             cmake_defs += [self.define("DYNAMIC_ARCH", "ON")]
         if self.spec.satisfies("platform=windows"):
@@ -611,11 +616,3 @@ class CMakeBuilder(spack.build_systems.cmake.CMakeBuilder):
             cmake_defs += [self.define("USE_OPENMP", "OFF"), self.define("USE_THREAD", "OFF")]
 
         return cmake_defs
-
-    # TODO: constrain this to @:0.3.27 once @0.3.28 is released released and confirmed fixed
-    # https://github.com/OpenMathLib/OpenBLAS/pull/4840
-    @run_after("install", when="platform=darwin")
-    def darwin_install_name(self):
-        # The shared library is not installed correctly on Darwin; fix this
-        fix_darwin_install_name(self.prefix.lib)
-

--- a/var/spack/repos/builtin/packages/openblas/package.py
+++ b/var/spack/repos/builtin/packages/openblas/package.py
@@ -612,8 +612,7 @@ class CMakeBuilder(spack.build_systems.cmake.CMakeBuilder):
 
         return cmake_defs
 
-    @run_after("install")
-    @when("platform=darwin")
+    @run_after("install", when="platform=darwin")
     def darwin_install_name(self):
         # The shared library is not installed correctly on Darwin; fix this
         fix_darwin_install_name(self.prefix.lib)

